### ISSUE TITLE
Cann't enable feature "asm" of sha2-0.9.2 on aarch64

### ISSUE THREAD

**Issue by dgbo** [filecoin-project/lotus#5391]
Hi, all.

Recently, we find sha256 on AARCH64 degradated to `soft_compress` of `sha2-0.9.2` if `FIL_PROOFS_USE_MULTICORE_SDR = 1`.
We check the implementation of `soft_compress`, it does not use the accelerator, e.g. `sha256su0`, `sha256h`, available on AARCH64.
This causes serious performance issues of `Phase1` on AARCH64.

These accelerators are used when feature "asm" is enabled for `sha2-0.9.2`.
We tried to solve this by modifying dependencies of in `Cargo.toml` of `storage-proofs-porep`, like:
```
[target."cfg(target_arch = \"aarch64\")".dependencies]
sha2 = { version = "0.9.1", features = ["asm"] }
[target."cfg(target_arch = \"x86_64\")".dependencies]
sha2 = { version = "0.9.1", features = ["compress"] }
```

But when we recompile the lotus on a aarch64 server, the compress feature is still there, we didn't get a `asm` feature, so `soft_compress` is still used.
`rustc --crate-name sha2 --edition=2018 /root/.cargo/registry/src/github.com-1ecc6299db9ec823/sha2-0.9.2/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no **--cfg 'feature="compress"'** --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=8892b7c94e6a666f -C extra-filename=-8892b7c94e6a666f --out-dir /home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps -L dependency=/home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps --extern block_buffer=/home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps/libblock_buffer-468e1f0012c10418.rmeta --extern cfg_if=/home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps/libcfg_if-77deadd60c39a184.rmeta --extern digest=/home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps/libdigest-d3cd462a59060294.rmeta --extern opaque_debug=/home/dongbo/lotus/extern/filecoin-ffi/rust/target/release/deps/libopaque_debug-96a51e27e2ac78e4.rmeta --cap-lints allow --print native-static-libs -C target-feature=+sha2`

We also try to use `sha2 = { version = "0.9.1", features = ["asm"] }` no matter the under running platform is `x86` or `aarch64`.
Again the compress feature is still there, `soft_compress` is still used on aarch64.

We definitely missed something. Can anyone help here? Thanks a lot.

Best Regards.

**Comment by dignifiedquire** [comment]
- are you building with `RUSTFLAGS="target-cpu=native"`? 
- `sha2 = { version = "0.9.1", features = ["asm"] }` does not pin the version, you need to use `version = "=0.9.12"` if you want to do that

**Comment by dgbo** [comment]
> * are you building with `RUSTFLAGS="target-cpu=native"`?
Yes, it is built with `RUSTFLAGS="-C target-cpu=native -g" FFI_BUILD_FROM_SOURCE=1 make clean all`.

> * `sha2 = { version = "0.9.1", features = ["asm"] }` does not pin the version, you need to use `version = "=0.9.12"` if you want to do that
Tested, it still is `features = ["compress"]`.

With `sha2 = { version = "0.9.1", features = ["asm"] }`, we find that if only build `storage-proofs-porep-5.4.0`, we get `--cfg 'features="asm"'`.
When we build `storage-proofs-5.4.0`, `--cfg 'features="asm"'` is gone, the `--cfg 'features="compress"'` is added again.

We guess the `'features="asm"'` are overrided somewhere. We seach feature `compress` (`grep "compress" ./ -nr`) in `storage-proofs-5.4.0`, `storage-proofs-core-5.4.0', `storage-proofs-porep-5.4.0', `storage-proofs-post-5.4.0', seems this feature are only controlled by `Cargo.toml` in `storage-proofs-porep-5.4.0'. :(
